src/libusermode/running.cpp

```
@@ -0,0 +1,490 @@
+    trap->cb = hook_process_cb;
+    trap->data = new er_data_t(*er_data);
+    trap->breakpoint.lookup_type = LOOKUP_DTB;
+    trap->breakpoint.dtb = info->regs->cr3;
+    trap->breakpoint.addr_type = ADDR_VA;
+    trap->breakpoint.addr = rip;
+    if (!drakvuf_add_trap(drakvuf, trap))
+    {
+        delete (er_data_t*) trap->data;
+        delete trap;
+    }
+
+    drakvuf_remove_trap(drakvuf, info->trap, free_trap);
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+
+/**
+ * Sets hook on context switch as we cannot yet create a trap on
+ * target_process's func_name function. This is because the func_name function
+ * physical address might not yet be mapped and hence will require requesting
+ * page faults. Those can only be handled from the target_process context.
+ */
+void userhook::request_userhook_on_running_process(
+    drakvuf_t drakvuf,
+    addr_t target_process,
+    std::string dll_name,
+    std::string func_name,
+    callback_t cb,
+    void* extra)
+{
+    drakvuf_trap_t* trap = new drakvuf_trap_t();
+    trap->type = REGISTER;
+    trap->reg = CR3;
+    trap->cb = wait_for_target_process_cb;
+
+    // Find target process pid, so we don't have to calculate it every time in
+    // the wait_for_target_process_cb.
+    vmi_pid_t target_pid;
+    if (!drakvuf_get_process_pid(drakvuf, target_process, &target_pid))
+    {
+        delete trap;
+        return;
+    }
+
+    trap->data = new er_data_t(this, target_process, target_pid, dll_name, func_name, cb, extra);
+    if (!drakvuf_add_trap(drakvuf, trap))
+    {
+        free_trap(trap);
+        return;
+    }
+}
+
+
+/**
+ * This is just a wrapper over a userhook::request_userhook_on_running_process
+ * method.
+ */
+bool drakvuf_request_userhook_on_running_process(
+    drakvuf_t drakvuf,
+    addr_t target_process,
+    std::string dll_name,
+    std::string func_name,
+    callback_t cb,
+    void* extra)
+{
+    if (!instance || !instance->initialized)
+    {
+        return false;
```

Actually, I would do a `throw -1` here, as I'm 99% sure this would be undefined state.

---

done